backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceEventsHandler.java

```diff
@@ -0,0 +1,217 @@
+        }
+        this.context = context;
+        this.customAttributes = customAttributes;
+        this.backtraceHandlerThread = backtraceHandlerThread;
+        this.api = api;
+        this.submissionUrl = settings.getSubmissionUrl(urlPrefix);
+        this.timeBetweenRetriesMillis = settings.getTimeBetweenRetriesMillis();
+
+        long timeIntervalMillis = settings.getTimeIntervalMillis();
+
+        BacktraceAttributes backtraceAttributes = new BacktraceAttributes(context, null, null);
+        this.application = backtraceAttributes.getApplicationName();
+        this.appVersion = backtraceAttributes.getApplicationVersionOrEmpty();
+
+        if (timeIntervalMillis != 0) {
+            final BacktraceEventsHandler handler = this;
+            handler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    handler.send();
+                    handler.postDelayed(this, timeIntervalMillis);
+                }
+            }, timeIntervalMillis);
+        }
+    }
+
+    /**
+     * Number of events in the queue
+     */
+    public int getCount() {
+        return events.size();
+    }
+
+    public int getMaximumNumberOfEvents() {
+        return this.maximumNumberOfEvents;
+    }
+
+    public void setMaximumNumberOfEvents(int maximumNumberOfEvents) {
+        this.maximumNumberOfEvents = maximumNumberOfEvents;
+    }
+
+    public abstract void sendStartupEvent(String eventName);
+
+    public void send() {
+        if (events == null || events.size() == 0) {
+            return;
+        }
+        sendEvents(events);
+    }
+
+    protected abstract void sendEvents(ConcurrentLinkedDeque<T> events);
+
+    protected abstract void sendEventsPayload(EventsPayload<T> payload);
+
+    @Override
+    public void handleMessage(Message msg) {
+        final BacktraceHandlerInputEvents input = (BacktraceHandlerInputEvents) msg.obj;
+        EventsResult result = getEventsResult(input);
+
+        if (input.eventsOnServerResponseEventListener != null) {
+            BacktraceLogger.d(LOG_TAG, "Processing result using custom event");
+            input.eventsOnServerResponseEventListener.onEvent(result);
+        }
+
+        retrySendEvents(input, result.getStatusCode());
+    }
+
+    protected void onMaximumAttemptsReached(ConcurrentLinkedDeque<T> events) {
+        return;
+    }
+
+    protected Map<String, Object> getAttributes() {
+        BacktraceAttributes backtraceAttributes = new BacktraceAttributes(context, null, customAttributes);
+        Map<String, Object> attributes = backtraceAttributes.getAllBacktraceAttributes();
+
+        DeviceAttributesHelper deviceAttributesHelper = new DeviceAttributesHelper(context);
+        attributes.putAll(deviceAttributesHelper.getDeviceAttributes());
+
+        return attributes;
+    }
+
+    protected abstract EventsPayload<T> getEventsPayload();
```

Comment:
Both implementation returns the same list - maybe we should mark it as vitual method and allow unique/summed events to override it (if you think it makes sense) or just provide a default implementation for this method here?

Reply:
See `SummedEventsHandler.getEventsPayload()` and `UniqueEventsHandler.getEventsPayload()`, they have different behavior because SummedEvents handler needs to clear Summed Events queue and UniqueEvents handler needs to update UniqueEvent timestamp. Should we still provide a default implementation, or can we keep it as abstract @konraddysput ?

Reply:
both implementation have similar logic - we can always invoke a method from an abstract class when we want to have a basic implementation. What do you think?